### PR TITLE
Update the APIKey in the Forwarder using config.OnUpdate

### PR DESCRIPTION
**Pull request by dustmop**
### What does this PR do?

When the API Key is refreshed using the secret Component, respond to the `config.OnUpdate` event by updating the API Key in the forwarder and update its health state.

### Motivation

Part of the api key refresh initiative.

### Describe how to test/QA your changes

__Setup__

Create two Datadog API Keys, assign them to the environment variables:

```
export DD_API_KEY0=[first api key...]
export DD_API_KEY1=[second api key...]
```

Create the following script named `secret-script.py` in the working directory:

```
#!/usr/bin/env python
import datetime
import json
import os
import sys

def main():
  content = sys.stdin.read()
  obj = json.loads(content)
  handles = obj['secrets']
  result = {}
  for h in handles:
    if h == 'my_api_key':
      if not os.path.isfile('alt.cfg'):
        result[h] = {'value': os.environ.get('DD_API_KEY0')}
      else:
        result[h] = {'value': os.environ.get('DD_API_KEY1')}
      continue
    result[h] = {
      'value': 'decoded_%s' % h,
    }
  print(json.dumps(result))

  dt = datetime.datetime.now()
  fout = open('secret-script.log', 'a')
  fout.write('[%s] %s\n' % (dt, result))
  fout.close()

if __name__ == '__main__':
  main()
```

Set its permissions properly:
```
chmod 500 secret-script.py
```

Configure the secrets backend in your datadog.yml file, as well as the api_key:
```
api_key: ENC[my_api_key]
secret_backend_command: script-script.py 
```

Tail the log file to see that the secret decoder is invoked every 4 seconds:
```
tail -f secret-script.log
```

__Run the agent and check status__

1) Run your agent: `agent run`

2) In another terminal, get the status to verify that your first api key is seen by the forwarder: `agent status`

check both the "Endpoints" and "API Keys status" sections

```
=========
Endpoints
=========

  https://app.datadoghq.com - API Key ending with:
      - abcde

...

  API Keys status
  ===============
    API key ending with abcde: API Key valid

```

__Refresh the key__

1) Activate the switch to change to the second api key `touch alt.cfg` (in the same working directory)

2) Refresh the api key: `agent secrets refresh`

3) Get the status again, the api key should display the second api key: `agent status`

### CODE REVIEW

File: comp/forwarder/defaultforwarder/forwarder_health.go
Comment (dustmop) [inline comp/forwarder/defaultforwarder/forwarder_health.go:244]: Renaming this method to make it more obvious that it updates state; it doesn't merely calculate and return a bool.

File: comp/forwarder/defaultforwarder/forwarder_health.go
Comment (pgimalac) [inline comp/forwarder/defaultforwarder/forwarder_health.go:171]: **🔨 warning:** ‏Not entirely sure but I guess this is called concurrently with the usual forwarder work, so we might need some synchronization around `fh.keysPerAPIEndpoint` ? (ie. a `RWLock`)

File: comp/forwarder/defaultforwarder/forwarder_health.go
Comment (ogaca-dd) [inline comp/forwarder/defaultforwarder/forwarder_health.go:98]: Not directly related to this PR: What do you think about having the following API for notifying when an configuration option is updated?

```go
config.OnStringUpdate("api_key", func (new_key string) { // func is called only when "api_key" changed
    fh.updateAPIKey(new_key)
} 
```
which is equivalent to
```go
config.OnStringUpdate("api_key", fh.updateAPIKey)

```

File: comp/forwarder/defaultforwarder/forwarder_health.go
Comment (dustmop) [reply comp/forwarder/defaultforwarder/forwarder_health.go:171]: After you mentioned mutexs I realized that there's memory aliasing going on between this and the defaultForwarder. Solving it involves a non-trivial refactor, but I'm working on it today.

File: comp/forwarder/defaultforwarder/forwarder_health.go
Comment (dustmop) [reply comp/forwarder/defaultforwarder/forwarder_health.go:98]: I think that's a good idea. After using `OnUpdate` a few times I've realized it's always necessary to check the key to see if the expected key is being updated, and right now it's very easy to mess that up and forget to do it.

File: comp/forwarder/defaultforwarder/forwarder_health.go
Comment (dustmop) [reply comp/forwarder/defaultforwarder/forwarder_health.go]: Nevermind, this was simpler than I thought. Added a mutex around the `fh.keysPerAPIEndpoint` as you said, and added mutexes to the `DomainResolvers` so that they can update and get their API keys safely.

File: comp/forwarder/defaultforwarder/forwarder_health.go
Comment (hush-hush) [inline comp/forwarder/defaultforwarder/forwarder_health.go:163]: Could we have the domain forwarder not rely on the `forwarderHealth` to receive configuration notification ? I think the less coupled we have those moving part the better.
It looks like we can update the config notification to notify with the old and new value and each part could be independent, no ?

**Review by hush-hush — COMMENTED** [review]
Added a comment, but looks good overall :+1: .

File: comp/forwarder/defaultforwarder/forwarder_health.go
Comment (dustmop) [reply comp/forwarder/defaultforwarder/forwarder_health.go:163]: Yes, makes sense. I updated the code as described.

File: comp/forwarder/defaultforwarder/default_forwarder.go
Comment (pgimalac) [inline comp/forwarder/defaultforwarder/default_forwarder.go:467]: **🥜 nitpick:** ‏If this is only used by tests it should likely move to a test file

File: comp/forwarder/defaultforwarder/default_forwarder_test.go
Comment (pgimalac) [inline comp/forwarder/defaultforwarder/default_forwarder_test.go:7]: **🥜 nitpick:** ‏fyi you don't need the test build tag in `_test.go` files
```suggestion
```

File: comp/forwarder/defaultforwarder/forwarder_health.go
Comment (pgimalac) [inline comp/forwarder/defaultforwarder/forwarder_health.go:259]: **🥜 nitpick:** ‏tiniest nitpick ever: you can move the `make` outside the `Lock`

File: comp/forwarder/defaultforwarder/resolver/domain_resolver.go
Comment (pgimalac) [inline comp/forwarder/defaultforwarder/resolver/domain_resolver.go:1]: ❓**Question**: `SingleDomainResolver` and `MultiDomainResolver` 's implementations seem to be very similar (eg. `UpdateAPIKey`), maybe their implementation could be factored ?
Not necessarily for this PR 😄

**Review by pgimalac — APPROVED** [review]
LGTM, only some style comments
Maybe still wait on the reviews from other people from the team who commented 😄

File: comp/forwarder/defaultforwarder/default_forwarder_test.go
Comment (GustavoCaso) [inline comp/forwarder/defaultforwarder/default_forwarder_test.go:26]: Could we use the config component?

File: comp/forwarder/defaultforwarder/resolver/domain_resolver.go
Comment (GustavoCaso) [reply comp/forwarder/defaultforwarder/resolver/domain_resolver.go:1]: Agree with that 😄

File: comp/forwarder/defaultforwarder/forwarder_health.go
Comment (GustavoCaso) [inline comp/forwarder/defaultforwarder/forwarder_health.go:197]: Are we actually ever using `apiKeyRemove` in `setAPIKeyStatus` ?

**Review by paulcacheux — APPROVED** [review]

**Comment by GustavoCaso** [comment]
❓ In the `validateAPIKey` function we create an http client for every request we made to the backend:

```go
func (fh *forwarderHealth) validateAPIKey(apiKey, domain string) (bool, error) {
	if apiKey == fakeAPIKey {
		fh.setAPIKeyStatus(apiKey, domain, &apiKeyFake)
		return true, nil
	}

	url := fmt.Sprintf("%s%s?api_key=%s", domain, endpoints.V1ValidateEndpoint, apiKey)

	transport := httputils.CreateHTTPTransport(fh.config)

	client := &http.Client{
		Transport: transport,
		Timeout:   fh.timeout,
	}
.
.
.
```

Would we benefit from reusing the client rather than creating a new client per request? Since we call `validateAPIKey` in the `healthCheckLoop` we might get better performance

File: comp/forwarder/defaultforwarder/default_forwarder.go
Comment (dustmop) [reply comp/forwarder/defaultforwarder/default_forwarder.go]: Good call, I like it!

File: comp/forwarder/defaultforwarder/default_forwarder.go
Comment (dustmop) [reply comp/forwarder/defaultforwarder/default_forwarder.go]: Ok, done

File: comp/forwarder/defaultforwarder/resolver/domain_resolver.go
Comment (dustmop) [reply comp/forwarder/defaultforwarder/resolver/domain_resolver.go:1]: Agreed that they're very similar, but doing it in another PR sounds like the best bet.

File: comp/forwarder/defaultforwarder/default_forwarder_test.go
Comment (dustmop) [reply comp/forwarder/defaultforwarder/default_forwarder_test.go]: Yes, turns out we can, much better.

File: comp/forwarder/defaultforwarder/forwarder_health.go
Comment (dustmop) [reply comp/forwarder/defaultforwarder/forwarder_health.go:199]: Good catch! The usage of it got removed in a previous commit, I've put it back now. It's necessary so that when the API Key is updated to use a new one, the old messages about the validity of the previous key get cleared.

**Comment by dustmop** [comment]
> Would we benefit from reusing the client rather than creating a new client per request? Since we call `validateAPIKey` in the `healthCheckLoop` we might get better performance

The `healthCheckLoop` only runs once every hour, so performance isn't a concern there. It's true that `checkValidAPIKey` can potentially call `validateAPIKey` multiple times, but that's only the case if the agent is using additional endpoints. In nearly all cases, there will only be 1 call to `validateAPIKey`. I'd rather keep the code simple and handle all http request construction in a single place.

**Review by hush-hush — APPROVED** [review]
Looks great :+1:

**Review by dineshg13 — APPROVED** [review]

**Review by rayz — APPROVED** [review]

**Pull request merged**
